src/test/webServer.ts

```diff
@@ -0,0 +1,144 @@
+import { launchProcess } from '../server/processLauncher';
+
+const DEFAULT_ENVIRONMENT_VARIABLES = {
+  'BROWSER': 'none', // Disable that create-react-app will open the page in the browser
+};
+
+const newProcessLogPrefixer = () => new stream.Transform({
+  transform(this: stream.Transform, chunk: Buffer, encoding: string, callback: stream.TransformCallback) {
+    this.push(chunk.toString().split(os.EOL).map((line: string): string => line ? `[WebServer] ${line}` : line).join(os.EOL));
+    callback();
+  },
+});
+
+export class WebServer {
+  private _killProcess?: () => Promise<void>;
+  private _processExitedPromise!: Promise<any>;
+  constructor(private readonly config: WebServerConfig) { }
+
+  public static async create(config: WebServerConfig): Promise<WebServer> {
+    const webServer = new WebServer(config);
+    if (config.port)
+      await webServer._verifyFreePort(config.port);
+    try {
+      const port = await webServer._startWebServer();
+      await webServer._waitForAvailability(port);
+      const baseURL = `http://localhost:${port}`;
+      process.env.PLAYWRIGHT_TEST_BASE_URL = baseURL;
+      console.log(`Using WebServer at '${baseURL}'.`);
+      return webServer;
+    } catch (error) {
+      await webServer.kill();
+      throw error;
+    }
+  }
+
+  private async _verifyFreePort(port: number) {
+    const cancellationToken = { canceled: false };
+    const portIsUsed = await Promise.race([
+      new Promise(resolve => setTimeout(() => resolve(false), 100)),
+      waitForSocket(port, 100, cancellationToken),
+    ]);
+    cancellationToken.canceled = true;
+    if (portIsUsed)
+      throw new Error(`Port ${port} is used, make sure that nothing is running on the port`);
+  }
+
+  private async _startWebServer(): Promise<number> {
+    let collectPortResolve = (port: number) => { };
+    const collectPortPromise = new Promise<number>(resolve => collectPortResolve = resolve);
+    function collectPort(data: Buffer) {
+      const regExp = /http:\/\/localhost:(\d+)/.exec(data.toString());
+      if (regExp)
+        collectPortResolve(parseInt(regExp[1], 10));
+    }
+
+    let processExitedReject = (error: Error) => { };
+    this._processExitedPromise = new Promise((_, reject) => processExitedReject = reject);
+
+    console.log(`Starting WebServer with '${this.config.command}'...`);
+    const { launchedProcess, kill } = await launchProcess({
+      command: this.config.command,
+      env: {
+        ...DEFAULT_ENVIRONMENT_VARIABLES,
+        ...process.env,
+        ...this.config.env,
+      },
+      cwd: this.config.cwd,
+      stdio: 'stdin',
+      shell: true,
+      attemptToGracefullyClose: async () => {},
+      log: () => {},
+      onExit: code => processExitedReject(new Error(`WebServer was not able to start. Exit code: ${code}`)),
+      tempDirectories: [],
+    });
+    this._killProcess = kill;
+
+    launchedProcess.stderr.pipe(newProcessLogPrefixer()).pipe(process.stderr);
```

Comment:
I think we need to drain `launchedProcess.stdout` just in case, otherwise it may block on Windows because stdio buffer will be full.